### PR TITLE
✉️ Add anyReasonableAddress()

### PR DESCRIPTION
**Pull request by MaciejNadolski98**

### CODE REVIEW

File: spec/truefi2/BorrowingMutex.spec [hunk 1]
```diff
@@ -23,9 +30,8 @@ rule functionDoesNotBanUnlockedBorrower() {
     require isReasonableEnv(e);
```
Comment (MaciejNadolski98) [inline spec/truefi2/BorrowingMutex.spec:30]: This stays as is, as returning structs in functions is not yet supported by Certora.

File: spec/truefi2/BorrowingMutex.spec [hunk 2]
```diff
@@ -6,6 +6,13 @@ methods {
 definition isReasonableAddress(address alice) returns bool = alice != 0 && alice != 1;
```
Comment (MaciejNadolski98) [inline spec/truefi2/BorrowingMutex.spec:6]: https://prover.certora.com/output/67403/e3982ef23f3c78f6850c?anonymousKey=fbc7c8d8088c64ca61e6b4ca70427d471381f601

File: spec/truefi2/BorrowingMutex.spec [hunk 3]
```diff
@@ -6,6 +6,13 @@ methods {
...
 definition isReasonableEnv(env e) returns bool = isReasonableAddress(e.msg.sender);
 
+function anyReasonableAddress() returns address {
+    address alice;
+    havoc alice;
```
Comment (yuchenlintt) [inline spec/truefi2/BorrowingMutex.spec:11]: Do we need this `havoc alice;` line?

If my understanding is correct, the declaration `address alice;` already quantifies over all possible `alice`s. `havoc` is only used to forget existing constraints on a variable's values:
https://certora.atlassian.net/wiki/spaces/CPD/pages/41156829/Hooks#Havoc-Assuming

**Review by yuchenlintt — CHANGES_REQUESTED** [review]

File: spec/truefi2/BorrowingMutex.spec (hunk 3 above)
Comment (yuchenlintt) [reply spec/truefi2/BorrowingMutex.spec:11]: It's still unclear to me whether multiple invocations of `anyReasonableAddress()` will return the same `alice` or a newly quantified one each time. Perhaps that's where `havoc` might be relevant?

Comment (MaciejNadolski98) [reply spec/truefi2/BorrowingMutex.spec:11]: Yep, that's exactly what this `havoc` is for. Tested by running a rule:
```
rule deterministic {
    assert anyReasonableAddress() == anyReasonableAddress();
}
```
Without `havoc` this rule was not violated.

Comment (skalermo) [reply spec/truefi2/BorrowingMutex.spec:11]: Just curious: would this `deterministic` rule also be violated if one of the addresses were not havoced?

Comment (yuchenlintt) [reply spec/truefi2/BorrowingMutex.spec:11]: In my understanding that should be the case! You wouldn't even need havoc. Any two different declarations would violate `alice == bob`.

Comment (MaciejNadolski98) [reply spec/truefi2/BorrowingMutex.spec:11]: @skalermo I tried running an example to answer your question by implementing a following function:
```
function anyBool(bool havocIt) returns bool {
    bool ret;
    if (havocIt) {
        havoc ret assuming ret@new != ret@old;
        return ret;
    } else {
        return ret;
    }
}
```
This, however, fails with an error message:
```
Syntax error in spec file (BorrowingMutex.spec:11:5): Variable ret is defined on the then branch and used on the else branch - this is disallowed
```
my understanding of this is that `havoc`ing is disallowed in multiple execution branch scenario. I'm still trying to figure out what are the exact restrictions on the keyword.

**Pull request closed without merging**